exposed/src/main/kotlin/org/jetbrains/exposed/sql/transactions/ThreadLocalTransactionManager.kt

```diff
@@ -7,138 +7,222 @@ import org.jetbrains.exposed.sql.Transaction
+            connection.setSavepoint("POINT_$name")
+        } else null
+
 
         override fun commit() {
-            if (connectionLazy.isInitialized())
-                connection.commit()
+            if (connectionLazy.isInitialized()) {
+                if (topLevelTransaction) {
+                    connection.commit()
+                } else {
+                    // do nothing in nested. close() will commit everything and release savepoint.
+                }
+            }
         }
 
         override fun rollback() {
             if (connectionLazy.isInitialized() && !connection.isClosed) {
-                connection.rollback()
+                if (topLevelTransaction) {
+                    connection.rollback()
+                } else {
+                    connection.rollback(savepoint)
+                    savepoint = connection.setSavepoint("POINT_$name")
+                }
             }
         }
 
         override fun close() {
             try {
-                if (connectionLazy.isInitialized()) connection.close()
+                if (topLevelTransaction) {
+                    if (connectionLazy.isInitialized()) connection.close()
+                } else {
+                    savepoint.let {
+                        connection.releaseSavepoint(it)
+                        savepoint = null
+                    }
+                }
             } finally {
                 threadLocal.set(outerTransaction)
             }
         }
 
+        private val name: String
+            get() {
+                var p: Transaction? = outerTransaction
+                var value: String = ""
+
+                while (p != null) {
+                    value = p.hashCode().toString(16) + "_" + value
+                    p = p.outerTransaction
+                }
+                return value + this.hashCode().toString(16)
+            }
     }
 }
 
-fun <T> transaction(db: Database? = null, statement: Transaction.() -> T): T = transaction(TransactionManager.manager.defaultIsolationLevel, TransactionManager.manager.defaultRepetitionAttempts, db, statement)
+fun <T> transaction(db: Database? = null, statement: Transaction.() -> T): T =
+    transaction(db.transactionManager.defaultIsolationLevel, db.transactionManager.defaultRepetitionAttempts, db, statement)
 
-fun <T> transaction(transactionIsolation: Int, repetitionAttempts: Int, db: Database? = null, statement: Transaction.() -> T): T {
+fun <T> transaction(transactionIsolation: Int, repetitionAttempts: Int, db: Database? = null, statement: Transaction.() -> T): T = keepAndRestoreTransactionRefAfterRun(db) {
     val outer = TransactionManager.currentOrNull()
 
-    return if (outer != null && (db == null || outer.db == db)) {
-        outer.statement()
+    if (outer != null && (db == null || outer.db == db)) {
+        val outerManager = outer.db.transactionManager
+
+        db?.let { db.transactionManager.let { m -> TransactionManager.resetCurrent(m) } }
```

Comment:
This line looks strange as you trying to reset manager for same database which should be also the same